cve_bin_tool/version_compare.py

```diff
@@ -38,66 +39,24 @@ def parse_version(version_string: str):
         raise UnknownVersion(f"version string = {version_string}")
 
     versionString = version_string.strip()
-    versionArray = []
 
     # convert all non alpha-numeric characters to be treated like . below
     # we could switch to a re split but it seems to leave blanks so this is less hassle
-    versionString = re.sub("[^0-9a-zA-Z]+", ".", versionString)
-
     # Note: This expression may need improvement if we need to handle unicode
+    versionString = re.sub("[^0-9a-zA-Z]+", ".", versionString)
 
-    # remove any trailing . then split
-    versionString = versionString.strip(".")
-    split_version = versionString.split(".")
-
-    # if the whole string was numeric then we're done and you can move on
-    if versionString.isnumeric():
-        versionArray = split_version
-        return versionArray
-
-    # Go through and split up anything like 6a in to 6 and a
-    number_letter = re.compile("^([0-9]+)([a-zA-Z]+)$")
-    letter_number = re.compile("^([a-zA-Z]+)([0-9]+)$")
-    for section in split_version:
-        # if it's all letters or all numbers, just add it to the array
-        if section.isnumeric() or section.isalpha():
-            versionArray.append(section)
-
-        # if it looks like 42a split out the letters and numbers
-        # We will treat 42a as coming *after* version 42.
-        elif re.match(number_letter, section):
-            result = re.findall(number_letter, section)
-
-            # We're expecting a result that looks like [("42", "a")] but let's verify
-            # and then add it to the array
-            if len(result) == 1 and len(result[0]) == 2:
-                versionArray.append(result[0][0])
-                versionArray.append(result[0][1])
-            else:
-                raise CannotParseVersionException(f"version string = {versionString}")
-
-        # if it looks like rc1 or dev7 we'll leave it together as it may be some kind of pre-release
-        # and we'll probably want to handle it specially in the compare.
-        # We need to threat 42dev7 as coming *before* version 42.
-        elif re.match(letter_number, section):
-            versionArray.append(section)
-
-        # It's not a "pure" alpha or number string, it's not something like rc12 or 44g
+    # if anything looks like a hash, we'll replace it with the word HASH and
+    # treat it like a pre-release in comparisons
+    versionString = re.sub("[a-fA-F0-9]{8,}", "HASH", versionString)
```

Comment:
This is so difficult... This now means that if somebody decides to create a version string with DDMMYYYY at the beginning, that's a hash. Is that a problem? Or, could that be a problem? Would it be worth it to go up to 12-14 characters instead?

Reply:
Hm, that's a good point.  So far I've seen dates only with delimiters, but that's in no way a guarantee that this is how they look across all the data.  Time to go grump my way through the entire database for a bit, I think...